Actor DSL

thedrow opened the pull request:
This PR contains the full implementation of the Actor DSL.

thedrow committed:
Begin work on actor.

thedrow committed:
All unskipped tests are passing.

thedrow committed:
Sort imports.

thedrow committed:
Remove print statement.

thedrow committed:
Move mock to dev dependencies.

thedrow committed:
Refactor state transition tests.

thedrow committed:
Support specifying timeout for resource acquisition.

thedrow committed:
Cleanup.

thedrow committed:
Drop curio support.

thedrow committed:
Refactor tests.

thedrow committed:
Resources are now accessible through self after acquisition.

thedrow committed:
Bump transitions-anyio to 0.2.0.

thedrow committed:
Remove assert.

thedrow committed:
NotAResourceError is now correctly raised when a timeout is specified as well.

thedrow committed:
Don't directly expose the exit stack.

thedrow committed:
Specify more types.

thedrow committed:
Add TODO comment about encapsulation violation.

thedrow committed:
Encapsulate resource lifecycle management in the actor class.

thedrow committed:
Resources can now be skipped if they are unavailable.

thedrow committed:
We can transition to the crashed state from any state.

thedrow committed:
We can transition to the starting state from the stopped state.

ahopkins committed:
Add test for transition back to starting after stopped

thedrow committed:
Implemented the ThreadedContextManager to support synchronous resources.

Fixes #13.

thedrow committed:
Sort imports.

thedrow committed:
Include ThreadedContextManager in __all__.

thedrow committed:
Implement synchronous resources.

thedrow committed:
Restore inheritance of HierarchicalAnyIOMachine.

thedrow committed:
Automatically transition from the starting state to the next state when the actor is stopped.

Fixes #20.

thedrow committed:
Fix formatting.

ghost commented:
## DeepCode's analysis on [#1f7746](https://www.deepcode.ai/app/gh/celery/jumpstarter/43dbf34e5e90ac53aff3acb524b0a47d115bb7c2/celery/jumpstarter/1f774693626a1ec9f352747b5d1282a267bc9506/pr/_/%2F/code/?utm_source=gh_review&c=0&w=1&i=1&) found:

- :warning: **1** warning, :information_source: **1** minor issue. :point_down:
## Top issues

<table>
  <thead>
    <tr>
      <th align="left">Description</th>
      <th align="left">Example fixes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td width="77%">First argument to super() should be the enclosing class ParallelGraphMachine Occurrences: <ul><li><a href="https://github.com/celery/jumpstarter/blob/1f774693626a1ec9f352747b5d1282a267bc9506/jumpstarter/states.py#L126">states.py:126</a></li></ul></td>
      <td width="23%"><a href="https://www.deepcode.ai/app/gh/celery/jumpstarter/43dbf34e5e90ac53aff3acb524b0a47d115bb7c2/celery/jumpstarter/1f774693626a1ec9f352747b5d1282a267bc9506/pr/_/%2Fjumpstarter%2Fstates.py/python%2Fdc%2Fsuper~first~argument~should~be~the~enclosing~class/code/?utm_source=gh_review&">:wrench: Example fixes</a></td>
    </tr>
    <tr>
      <td width="77%">The call to next should be guarded with a try/except block Occurrences: <ul><li><a href="https://github.com/celery/jumpstarter/blob/1f774693626a1ec9f352747b5d1282a267bc9506/jumpstarter/actors.py#L60">actors.py:60</a></li></ul></td>
      <td width="23%"><a href="https://www.deepcode.ai/app/gh/celery/jumpstarter/43dbf34e5e90ac53aff3acb524b0a47d115bb7c2/celery/jumpstarter/1f774693626a1ec9f352747b5d1282a267bc9506/pr/_/%2Fjumpstarter%2Factors.py/python%2Fdc%2Funguarded~next~call/code/?utm_source=gh_review&">:wrench: Example fixes</a></td>
    </tr>
  </tbody>
</table>


#### 👉 View analysis in [**DeepCode’s Dashboard**](https://www.deepcode.ai/app/gh/celery/jumpstarter/43dbf34e5e90ac53aff3acb524b0a47d115bb7c2/celery/jumpstarter/1f774693626a1ec9f352747b5d1282a267bc9506/pr/_/%2F/code/?utm_source=gh_review&c=0&w=1&i=1&) | [_Configure the bot_](https://www.deepcode.ai/app/gh/?ownerconfig=celery)

thedrow committed:
Fix import.

thedrow committed:
Install dev dependencies as well.

thedrow committed:
Add backport for classproperty

thedrow committed:
Tests are now backwards compatibile with Python 3.7.

thedrow committed:
Update poetry.lock.

thedrow committed:
Refactor importing.

thedrow committed:
Export graphs if pygraphviz is installed.

auvipy commented on jumpstarter/resources.py:
```diff
@@ -0,0 +1,119 @@
+
+class NotAResourceError(Exception):
+    def __init__(self, resource_name: str, return_value):
+        super(NotAResourceError, self).__init__(
```
is explicit call to class needed here inside super??

thedrow replied:
No, that's how pycharm autocompletes it. I'll fix this.

auvipy commented on jumpstarter/backports.py:
```diff
@@ -0,0 +1,9 @@
+class classproperty(property):
+    def __get__(self, obj, objtype=None):
+        return super(classproperty, self).__get__(objtype)
```
calling explicitly or autocomplete? :D

thedrow committed:
Add a way to automatically add type annotations by running the tests.

thedrow committed:
Remove explict super() since it is not needed in Python 3.

thedrow committed:
Install libgrahviz-dev in CI.

thedrow committed:
Use xdist to parallelize tests.

thedrow committed:
Remove unused imports and variables.

thedrow committed:
Upgrade code to Python 3.7 and above when formatting.

thedrow committed:
Generate a unique identifier for each actor instance.

thedrow committed:
Merge pull request #16 from celery/sync-resources

Implement synchronous resources

thedrow committed:
Reorganize tests.

thedrow committed:
Customize actor ID.

thedrow committed:
Register cancel scope as a resource now that resources are implemented.

thedrow committed:
Prepare for health-checks.

thedrow committed:
pytest will now annotate test failures in CI.

thedrow committed:
Workaround PyCharm's debugger.

thedrow committed:
Add capability to pause and resume the actor without stopping it completely.

Note that the build is failing due to a bug in transitions. It is fixed in https://github.com/pytransitions/transitions/pull/506.

thedrow committed:
Add transitions between actor health states.

thedrow committed:
Add colors to the state machine graph and change layout.

thedrow committed:
Use sentinel value.

thedrow committed:
Initial draft of actor inheritance.

thedrow committed:
No idea why initial nested states don't work anymore but in any case, this is functionally equivalent.

thedrow committed:
Added comment.

thedrow committed:
Once more with tests.

thedrow committed:
Add the initialize() trigger.

thedrow committed:
Add color to state machine graph edges.

thedrow committed:
Add typing information.

thedrow committed:
Create a transition from crashed to starting.

thedrow committed:
Set state machine name for logging purposes.

thedrow committed:
Basic API for parallel states.

thedrow committed:
Return all parallel states if they are available.

thedrow committed:
Merge branch 'actor' into actor-inheritance

thedrow committed:
Measure test coverage.

thedrow committed:
Use NestedAsyncTransition to trigger nested initial states.

Also specify initial nested states using ChildStateEnum, a special dictionary which allows to use the children of the Enum as attributes like a regular Enum.

thedrow committed:
Merge pull request #27 from celery/actor-inheritance

Implementation of Actor Inheritance

thedrow committed:
Merge branch 'main' into actor

thedrow committed:
Begin work on restarting the actor.

Achieving parallel hierarchical async state machine is not the easiest feat, but it's certainly doable if we reduce it to a very private case such as ours.

thedrow committed:
I'll sort out the tests later but this is generally the idea.
You can't restart since the condition refers to the main state machine.
If you can, that means that you're in an allowed state and thus enter two states in parallel.

thedrow committed:
Raise a machine error when you cannot transition from a state to restarting.

thedrow committed:
During restart, we stop the actor and then start it.
Not the other way around...

thedrow committed:
Add the model to all parallel state machines.

thedrow committed:
Add parallel state machine support even if diagram support is not available.

thedrow committed:
Actor restart process is now implemented.

Currently, the actor is stuck when restarting from the crashed state.
We will investigate.

thedrow committed:
Refactor the test and introduce a timeout in case the test runs forever.

thedrow committed:
isort is a development dependency.

thedrow committed:
Fun with graphs.

We now display the parallel state machines and their subgraphs.

thedrow committed:
Add support for optionally setting bootup and shutdown events.

thedrow committed:
Progress to substates after transitioning from crashed to stopping or starting.

thedrow committed:
Extend allowed transitions into restart.

thedrow committed:
Format pyproject.toml automatically.

thedrow committed:
Treat warnings as errors.

Ignore a specific PytestDeprecationWarning since pytest-subtests is triggering it.

thedrow committed:
Add type annotations.

thedrow committed:
Actor can now restart.

test_restart_states needs to be refactored since the ActorStateMachine is now almost completely decoupled from the ActorRestartStateMachine.

thedrow committed:
Refactored tests.

thedrow committed:
Generate API reference.

thedrow committed:
Publish public API through the main package.

thedrow committed:
Sidebar now points to the right repository.

thedrow committed:
Parse type annotations and use them in the documentation.

thedrow committed:
Avoid using nox_poetry deprecated API.

thedrow committed:
Don't install all extras unless asked to.

thedrow committed:
Merge pull request #28 from celery/actor-restart

Actor Restart Process

thedrow committed:
Add types.

thedrow committed:
Sort imports.

thedrow committed:
Skip test if diagrams are not supported due to missing extra dependency.

thedrow committed:
Ensure DeepCode ignores what git ignores.

thedrow committed:
Restore CI runs.

thedrow committed:
Reformat.

thedrow committed:
Cache dependencies from PyPi.

thedrow committed:
Install graphviz as well.

thedrow committed:
Run all options in parallel.

thedrow committed:
Skip graphviz if possible.

thedrow committed:
Bump minimum transitions version.

thedrow committed:
Merge branch 'main' into actor

restyled-commits committed:
Fixed formatting (whitespace)

restyled-commits committed:
Fixed formatting (isort)

restyled-commits committed:
Fixed formatting (black)

thedrow committed:
Configure isort to be compatible with black.

thedrow committed:
Use classproperty backport.

thedrow committed:
Add networkx as a dependency.

thedrow committed:
Allow actors to specify dependencies.

thedrow committed:
Start/stop dependencies.

Currently, the transitions are not added to the list of callbacks for some reason.

thedrow committed:
Start/stop dependencies.

restyled-commits committed:
Fixed formatting (isort)

restyled-commits committed:
Fixed formatting (black)

thedrow committed:
Return dependencies' state as well.

thedrow committed:
Cache poetry dependencies.

thedrow committed:
Split to subtests.

thedrow committed:
More test coverage.

thedrow committed:
Merge pull request #33 from celery/actor-dependencies

Actor dependencies

thedrow committed:
Add regions for code folding.

thedrow committed:
Run unit tests only if there has been a meaningful change.

Run the unit tests if any Python file has changed, the dependencies have changed or if the CI pipeline itself has changed.

thedrow committed:
Merge pull request #34 from celery/run-ci-when-needed

Run unit tests only if there has been a meaningful change

thedrow committed:
Update dependencies.

thedrow committed:
Merge branch 'main' into actor

thedrow committed:
Include other sphinx extensions in extra.

thedrow committed:
Add extras for event loops.

thedrow committed:
Add uvloop to the testing matrix.

thedrow committed:
Avoid reinstalling non-dev dependencies.

restyled-commits committed:
Fixed formatting (black)

thedrow committed:
Update dependencies.

thedrow committed:
Update dependencies.

thedrow committed:
Bump transitions and transitions-io.

thedrow committed:
Bump minimum Python version.

thedrow committed:
Bump sphinx-pyreverse.

thedrow committed:
Bump pytest plugins.

thedrow committed:
Migrate code to anyio 3.x.

restyled-commits committed:
Fixed formatting (black)

Commit:
Update dependencies

restyled-commits committed:
Fixed formatting (black)

Commit:
Fix syntax

Commit:
Fix noxfile

Commit:
Use trio from pypi

Commit:
Update uvloop
